### DISCUSSION TITLE
First SSR Page request is too slow.

### DISCUSSION THREAD

**Discussion by molink-huseong** [vercel/next.js#38012] [Help]
Hi, I'm adding getServerSideProps function in my page and after that suffering slow rendering issue.
The issue only happen when entering the page which using ssr, and also only first time.

I captured the request using chrome developer tool network tab and found that ''waiting (TTFB)" takes 12 seconds compare to content download as 1ms.
The initiator is router.js and size is only 465 Byte. I can't understand why this request takes so long time
![capture](https://user-images.githubusercontent.com/97883121/175777098-ff3d051b-a19c-48ab-a6b4-88c3f822a841.PNG)

**Comment by wereHamster** [comment]
What's the getServerSideProps function doing? Is it loading data from a remote system, database etc, something that could take those 12+ seconds?

**Reply by molink-huseong** [reply]
Hi,
My getServerSideProps call one API which normally takes 50ms.
For test, I changed my getServerSideProps return empty object, no API call.
But first routing also takes 12 seconds although getServerSideProps doesn't make API call.
I added logging at getServerSideProps and also find that getServerSideProps call after 12 seconds.

Summarize-Timeline
0s Route.Push Call
pending state route.js
12s getServerSideProps Call
12.3s New page first rendering

Summarize-Case
Page A, Page B

Enter Page B(with getServerSideProps)
Takes 1sec

Enter Page A, and move to Page B(with getServerSideProps)
Takes 12sec

Enter Page B(with getServerSideProps) => Page A => Page B (with getServerSideProps)
Last routing Takes 1sec

Enter Page A => Page B(with getServerSideProps) => Page A => Page B(with getServerSideProps)
Last routing Takes 1sec

Enter Page A => Page B(with getServerSideProps) => Page A => (5 minutes after) => Page B(with getServerSideProps)
Last routing Takes 12sec

Enter PageA => Page B(without getServerSideProps)
Takes 1sec

**Reply by wereHamster** [reply]
Is that in production or development?

**Reply by molink-huseong** [reply]
both, production and development.

**Reply by wereHamster** [reply]
getServerSideProps taking longer in development is common, because the server has to compile the code first. And the observation that it is faster once the code is compiled and takes longer if it's the first time, or because the compiled code was thrown out of the cache (after a few minutes) all point that you are still running in dev mode. Are you absolutely sure that this happens in production (when the server runs via `next build && next start`)?

**Reply by molink-huseong** [reply]
Yes! I'm absolutely sure. I checked in production and development env (deployed by vercel).
Prod (vercel), Dev (vercel), Local all enviroment happen..
You can check in my website ( https://molink.life ).
You can ignore the language and press any block to move.

**Reply by wereHamster** [reply]
You could try using the [Server Timing API](https://web.dev/custom-metrics/#server-timing-api) to add to the response where exactly time is being spent.

**Reply by molink-huseong** [reply]
Thanks, I will use it and find the reason.

**Comment by molink-huseong** [comment]
I find the element size of Page B(with getServerSideProps) effect response time.
I guess adding getServerSideProps changes pre-rendering to ssr on each request.
But I can't understand why only route.push make delay.
'Enter Page B' and 'Move to Page B from Page A' looks similar on client side, but takes '1sec' and '12sec'

**Reply by Raheel-Saleem** [reply]
@molink-huseong  did you find the solution? I am also having the same problem.

**Comment by flex-jonghyen** [comment]
I have similar problem.
I use Argo CD to manage SSR server pod.
I config Argo CD to request health check to `/api/health` to SSR server.
While making SSR page at first time, SSR server can not response health check for about 3 seconds.
Now i solved to increase timeout to 10 seconds.
But i want to know what is blocking response and what happen while first SSR page request.

**Comment by flex-jonghyen** [comment]
I simulated Argo CD doing and profile Next server CPU.
I have CPU profile file. https://drive.google.com/file/d/1NWExIEAF17zK4vSscrs5ParaU68boezn/view?usp=sharing
Server spent almost time to read file of page and components is used on the page.
I spent so many time to solve this problem. Help me. 😭

**Comment by snordale** [comment]
Also having issues with this.

**Comment by mamadou-niakate** [comment]
I am also facing the same issue. I have my blog built with Next.js (SSR) and hosted on Amplify. The initial load take at least 10 seconds and the others requests are fast. What i dont understand is that: let's say we have 3 users **A**, **B** and **C**, when **A** try to access tothe site with his device, its first page loading is tooooo slow but for **B** and **C** from their devices the first page loading is very quick. Here is the site: [beingdev.online](http://www.beingdev.online/)

**Comment by abdulrahimiliasu** [comment]
I had similar problem some time ago, if you're using `getServerSideProps` in index file of your site (i.e first landing page) , then I recommend fetching whatever it is on client side and not server side, that way your components can load first before the data. Not a solution but a workaround.

**Comment by maxratajczak** [comment]
Yep, I had this issue for a looonggg time. It's so sad to see Vercel completely ignore it or act like it doesn't exist. For this reason I switched to host my app on digital ocean on an actual server rather than Vercel's serverless functions, and now my SSR pages load instantly. The problem is the way Vercel runs the SSR, they use serverless functions which have a cold boot time on first load. Vercel could easily fix this by adding a min amount of instances option and the cold boots would lower, but they haven't.

First few things you can do is get your initial load size as low as possible (you can see this when running 'next build'). Use NextJs bundle analyzer to see where your big chunks of code is and use dynamic import to code split. When it comes to what is run in getServerSizeProps, try to run as little code as possible, only get what is needed for a minimum viable page, and then fetch the rest of needed data when page already loads.

I am also looking into the different speeds of using a SDK websocket VS simple fetch api when getting data on SSR. I cannot confirm this is correct yet, but I'm thinking opening a web socket takes much longer on ever single request rather than using a rest API fetch. For example, I'm using Firebase firestore as my database and I'm only using their rest api on server side rendering rather than using the javascript sdk.

**Reply by Gbuomprisco** [reply]
Check out the `preferRest` option in Firestore - it will considerably improve cold boots

**Comment by denvudd** [comment]
I also have a similar problem. In my [id].tsx file, getServerSideProps function makes an API request using RTK Query (using next redux wrapper), and the data is passed to the component via props, after that data is used in the components. The initial load is slow (around 7 seconds, as you can see on the screenshot below). Subsequent loads of [id].tsx are faster, but after some time, the loading takes 7 seconds again. Honestly, I have no idea why this is happening.

The most interesting thing about this behavior is that in development mode, the initial load is not as slow as in production. I have already tried switching from Vercel to Netlify for deploy, but there is no difference.

By the way, if anyone knows: is it possible that "heavy" components in [id].tsx can slow down the initial load? I'm using UI Library Ant Design and they are kinda heavy.

My [id].tsx:
```
import React from "react";
import DetailLayout from "@/layouts/DetailsLayout";
import MovieDetailsBlock from "@/components/MovieDetailsBlock/MovieDetailsBlock";
import {
  getMovieDetails,
  getRunningQueriesThunk,
} from "@/redux/api/movies/slice";
import { wrapper } from "@/redux/store";
import { MovieDetails } from "@/redux/api/movies/types/MovieDetailsType";

interface MovieDetailsPageProps {
  id: number;
  data: MovieDetails;
}

export const getServerSideProps = wrapper.getServerSideProps(
  (store) => async (context) => {
    const { id } = context.query;

    if (typeof id === "string") {
      await store.dispatch(
        getMovieDetails.initiate({ id, params: "language=uk-UA&page=1" })
      );
    }

    const { data } = getMovieDetails.select({
      id,
      params: "language=uk-UA&page=1",
    })(store.getState());

    await Promise.all(store.dispatch(getRunningQueriesThunk()));

    return {
      props: { id, data },
    };
  }
);

const MovideDetailsPage: React.FC<MovieDetailsPageProps> = ({ id, data }) => {
  return (
    <>
      <DetailLayout>
        {data && <MovieDetailsBlock id={id} data={data} />}
      </DetailLayout>
    </>
  );
};

export default MovideDetailsPage;
```

Performance results (json file):
![image](https://github.com/vercel/next.js/assets/68691654/0a2a8380-a291-42f7-bda0-e5881d7a38e7)

**Comment by happydev000** [comment]
Do we have the proper solution for this yet?

**Reply by maxratajczak** [reply]
Nope, Vercel hasn't addressed it. My solution has been to use DigitalOceans app platform to host my app which uses an actual server to run SSR rather than Vercel's serverless functions. SSR loads are instant for me when using DigitalOcean.

**Reply by happydev000** [reply]
Understood. Currently, I am deploying nextjs 12 app in AWS. The problem is that I am facing the memory spike issue from the first page loading. My repo is monorepo and I am deploying thru AWS ECS. Any idea about this?

I increased the ec instance memory size to Large to resolve this issue but I am wondering if this is the correct solution for this because the first page loading is really slow.

**Reply by dlvx** [reply]
I just ran a quick test deploying my app to Digital Ocean and it certainly is much faster than my Vercel deployment. Lighthouse performace score went from 67 to 86, FCP from 2.1s to 1.3s and total blocking time from 390ms to 130ms. I'll probably be migrating in the next couple of days as this is a critical thing. I hope Vercel finds a way to address this as the service as a whole has been pretty good, nevertheless, I only have good things to say about Digital Ocean.

**Comment by baptistemanson** [comment]
We are facing a similar issue. The first request to a page is slow, then when it's hot, it runs well. Across users, so we know it's a vercel thing.

**Reply by MONTYYYYY** [reply]
Same here. Hope it gets fixed asap

**Comment by AlexanderHubar** [comment]
Same for me. I'm going to see if DigitalOcean really helps with that. I hope so. But if it does, it will be sad to switch to DigitalOcean from Vercel. I really like the Vercel ecosystem.

**Comment by sluxzer** [comment]
Still having the same issue
